pkg/kubelet/kubelet_pods.go

```diff
@@ -1020,6 +1021,69 @@ func (kl *Kubelet) removeOrphanedPodStatuses(pods []*v1.Pod, mirrorPods []*v1.Po
 	kl.statusManager.RemoveOrphanedStatuses(podUIDs)
 }
 
+// deleteMirrorPod uses mirror client to delete the mirror pod from the API server
+func (kl *Kubelet) deleteMirrorPod(podFullName string) {
+	deleted, err := kl.podManager.DeleteMirrorPod(podFullName, nil)
+	if err != nil {
+		klog.Errorf("encountered error when deleting mirror pod %q : %v", podFullName, err)
+		return
+	}
+	if !deleted {
+		klog.Errorf("mirror pod %q was not deleted", podFullName)
+		return
+	}
+	// the mirror pod has been removed from api server, clean the entry for bookkeeping
+	kl.podManager.MarkPodTerminated(podFullName)
+}
+
+// deleteOrphanedMirrorPods checks whether grace period for orphaned mirror pod has elapsed.
+// Once the grace period has passed, podManager.DeleteMirrorPod() is called to delete mirror pod
+// from the API server
+func (kl *Kubelet) deleteOrphanedMirrorPods() {
+	podFullNames := kl.podManager.GetOrphanedMirrorPodNames()
+	for _, podFullName := range podFullNames {
+		start := time.Now()
+		uidWithGracePeriod, ok := kl.podManager.GetUIDForPodPendingTermination(podFullName)
+		if !ok {
+			klog.Warningf("unable to find pod UID for %q, likely kubelet restarted", podFullName)
+			kl.deleteMirrorPod(podFullName)
+			continue
+		}
+		allContainersExited := true
+		if start.Before(uidWithGracePeriod.Till) {
```

Comment:
i agree that the kubelet restart scenario has no ideal solution, but i wonder if a simpler approach is the following.

could we avoid tracking this state information in pod manager by instead just saying "do not delete a pod if its present in the pod killing channel"?  The pod killing channel is the real thing you could still race against in this approach, and its the actual source of truth internally that the kubelet (pending a restart) still has work to do to remove that pod?


given that we have no checkpoint on pod manager, is it better for now to check if the pod exists in the pod killing channel?  if the pod exists in the pod killing channel, it implicitly means it has not yet been killed?

Reply:
Derek:
Thanks for the detailed notes and suggestion on simplifying the PR.
Here is my response w.r.t. suggestion on pod killing channel check.

Kubelet#podKiller reads from podKillingCh and calls killPod() on the pod to be killed.
It is unknown how long the pod killing request would stay in the podKillingCh.

My expectation w.r.t. grace period from pod Spec is that the mirror pod should be deleted after this amount of time has elapsed from the time pod manifest is removed. The grace period would become less meaningful if we use a different starting point in time because we may not have much control over the factors that lead to the delayed serving of the pod killing request.

I think in your suggestion above, you meant checking whether the killing set in Kubelet#podKiller contains the pod we are interested in:
```
	killing := sets.NewString()
```
(since I don't know, in golang, how to check whether some request is in the channel without taking it out of the channel and serving it).
Actually I thought about this and didn't choose this route because of the following consideration:

We need to distinguish the following two scenarios:

1. the pod killing request is in podKillingCh and has not been served by the go routine yet
2. the pod killing request has been served by the go routine and is no longer in the killing set:
```
			go func(apiPod *v1.Pod, runningPod *kubecontainer.Pod) {
```
Since the orphan pod cleaner runs periodically, it becomes tricky how to reliably distinguish between the two scenarios (without making the Kubelet#podKiller logic more complicated).

w.r.t. kubelet restart, in the current formation of the PR, we remove the mirror pod after kubelet restart if it is detected that the mere presence of the mirror pod was due to wait (in prior instantiation of the kubelet) for the grace period to expire.

Reply:
>Kubelet#podKiller reads from podKillingCh and calls killPod() on the pod to be killed.
>It is unknown how long the pod killing request would stay in the podKillingCh.
>Once the pod killing request is served, we don't know how long the killing of container would take.

The desire is to remove the mirror pod when we know its associated local pod resource is no longer running on the node.  It is right that we do not know how long the pod killing request would stay in the pod killing channel, but this is precisely the reason your code (which is right) gates deletion of the mirror pod until it sees all containers have reported exited.

The flow I see is the following:

1. the pod is removed from pod manager (i.e. its orphaned)
1. the pod is inserted in the pod killing channel (pending deletion)

if the sync loop invokes pod housekeeping before the pod killer acts on the pod in the channel, we will delete the mirror pod prematurely.  this is the core race condition we need to address.

The PR presented attempts to address this situation by making the pod manager aware of the terminating grace period for orphaned mirror pods.  

1. The housekeeping loop checks if the pod is still in its grace period from pod manager 
1. The housekeeping loop checks if the pod container status is terminal

The second check is implicitly ensuring that the pod killer channel has not acted on the pod yet.

Instead of changing the workflow around pod manager and the states it maintains, it seems like we should make the cleanup routine aware that a pod is orphaned AND know if the pod killer channel has completed its task before deleting the mirror pod.

What the changes in your code flow appear to do is try to make pod manager grace period aware, but its still disconnected from the content of the channel which is the actual condition we are racing against.

It's important we understand the sequence of actions from delete to removal from the API server which is why recording the state of interactions in sync loop, state of objects in pod manager, state of objects in pod killing channel are all important to resolving the core issue.

WDYT?

Reply:
The runningPods in Kubelet#HandlePodCleanups() are obtained from kl.runtimeCache, hence they can be slightly out of date.
Currently Kubelet#HandlePodCleanups() interacts with pod killer via sending pod to podKillingCh only.

Some implications of this are:
1. before unwanted pod is killed, Kubelet#HandlePodCleanups() would continue to add the pod to podKillingCh. pod killer would filter out the repetition using the killing set. This repetition seems unnecessary.
2. If we expose the killing set alone, it is still tricky for Kubelet#HandlePodCleanups() to tell whether the pod killer has finished the call to kl.killPod().

In the following sequence of events, T3 and T4 are very close (in the sense that runtimeCache isn't updated).

T1 Kubelet#HandlePodCleanups() finds pod 1 as unwanted, sends it to podKillingCh
T2 pod killer takes pod 1 out of podKillingCh and works on the call to kl.killPod()
T3 call to kl.killPod() call returns to pod killer, pod 1 is taken out of the killing set.
T4 Kubelet#HandlePodCleanups() gathers runningPods (including pod 1, according to runtimeCache) and desiredPods, performs the loop again (sending pod 1 to podKillingCh):
```
	for _, pod := range runningPods {
		if _, found := desiredPods[pod.ID]; !found {
```

If we look at the killing set at T4, we wouldn't know whether the pod termination has completed, without checking the pod status (via containerRuntime.GetPodStatus).

I thought about adding another Set, called toBeKilled, which contain pods in runningPods but not in desiredPods, for pod killer. This would get around the task of checking pod(s) currently in the podKillingCh, but seems inadequate for the scenario described above where the pod is neither in podKillingCh (toBeKilled set) nor in killing set.

Hence I chose to depend on the call to containerRuntime.GetPodStatus in Kubelet#deleteOrphanedMirrorPods which would tell us whether the pod has really terminated.

I will keep thinking about the interactions between the following :
* pod manager
* housekeeping (Kubelet#HandlePodCleanups)
* pod killer
* runtimeCache

thanks

Reply:
From my comment yesterday:
```
My expectation w.r.t. grace period from pod Spec is that the mirror pod should be deleted after this amount of time has elapsed from the time pod manifest is removed.
```
It is responsibility of Kubelet#HandlePodCleanups(Kubelet#deleteOrphanedMirrorPods, to be exact) to clean up orphaned mirror pods since the Kubelet#killPod() call may take longer than grace period to complete.
This means Kubelet#deleteOrphanedMirrorPods needs to know when the deletion of static pod is initiated so that it can keep track of the time before grace period expires.

I thought of introducing a map, podTerminationMap into Kubelet class which keeps the following information:
1 full pod name 
2 the timestamp when static pod deletion is initiated
3 the timestamp when the pod containers are terminated

#2 is used by Kubelet#deleteOrphanedMirrorPods to know when grace period expires.
#3 is filled by pod killer when Kubelet#killPod() returns

Kubelet#deleteOrphanedMirrorPods would iterate podTerminationMap and call mirror client to remove mirror pod when either of the following is true:
* grace period for the underlying pod has passed
* pod killer informs deleteOrphanedMirrorPods() via #3 that the pod has terminated

However we need to coordinate accesses between Kubelet#deleteOrphanedMirrorPods and pod killer to the map so that no map entry for the pod is leaked.

Please let me know if the above is simpler than the formation in this PR.

Reply:
Here was excerpt from kubelet log on a patched 1.18 master node (using patch in the above approach) where kube-apiserver-pod.yaml was moved out of manifest dir.
```
Jun 30 22:38:41 ip-10-0-136-93 hyperkube[768651]: I0630 22:38:41.942971  768651 kubelet_pods.go:1031] checking entry "kube-apiserver-ip-10-0-136-93.ec2.internal_openshift-kube-apiserver", 2020-06-30 22:39:45.76633046 +0000 UTC m=+394.295683069, 2020-06-30 22:38:41.166458121 +0000 UTC m=+329.695810607
Jun 30 22:38:41 ip-10-0-136-93 hyperkube[768651]: I0630 22:38:41.943004  768651 kubelet_pods.go:1040] deleting pod "kube-apiserver-ip-10-0-136-93.ec2.internal_openshift-kube-apiserver", 2020-06-30 22:38:41.166458121 +0000 UTC m=+329.695810607
Jun 30 22:38:41 ip-10-0-136-93 hyperkube[768651]: I0630 22:38:41.943018  768651 mirror_client.go:124] Deleting a mirror pod "kube-apiserver-ip-10-0-136-93.ec2.internal_openshift-kube-apiserver" (uid (*types.UID)(nil))
```
Please note that '2020-06-30 22:38:41.166458121' (in the first line) was the timestamp when pod killer completed the Kubelet#killPod call.
Also note that 2020-06-30 22:39:45.76633046 was the deadline for the grace period - the mirror pod was deleted prior to the deadline since the containers have been destroyed.